### PR TITLE
Fix/redundant network calls to notary during auth

### PR DESCRIPTION
**Pull request by annekebr**
Fixes #

## Checklist

- [x] PR is rebased to/aimed at branch `develop`
- [x] PR follows [Contributing Guide](https://github.com/sse-secure-systems/connaisseur/blob/master/docs/CONTRIBUTING.md)
- [x] Added tests (if necessary)
- [x] Extended README/Documentation (if necessary)
- [x] Adjusted versions of image and Helm chart in `Chart.yaml` (if necessary)

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -50,10 +50,9 @@ A simple starting point may be a minikube cluster with e.g. a [Docker Hub](https
 In case you make changes to the Connaisseur container image itself or code for that matter, you need to re-build the image and install it locally for testing.
 This requires a few steps:
 
-1. In `helm/values.yaml`, set `imagePullPolicy` to `IfNotPresent`.
-2. Configure your local environment to use the Kubernetes Docker daemon. In minikube, this can be done via `eval (minikube docker-env)`.
-3. Build the Connaisseur container image via `make docker`.
-4. Install Connaisseur as usual via `make install`.
+1. Configure your local environment to use the Kubernetes Docker daemon. In minikube, this can be done via `eval $(minikube docker-env)`.
+2. Build the Connaisseur container image via `make docker`.
+3. Install Connaisseur as usual via `make dev-install`.
```
Comment (Starkteetje) [inline docs/CONTRIBUTING.md:55]: as usual doesn't really work here

File: .github/actions/alerting-endpoint/action.yaml [hunk 2]
```diff
@@ -16,4 +16,4 @@ runs:
         SEARCH_PATH=.[0].NetworkSettings.Networks.${CONTAINER_NETWORK}.IPAddress
         IP=$(docker container inspect ${CONTAINER_ID} | jq -r ${SEARCH_PATH})
         echo IP=${IP} >> ${GITHUB_OUTPUT}
-      shell: bash
+      shell: bash
```
Comment (Starkteetje) [inline .github/actions/alerting-endpoint/action.yaml:19]: why this change?

File: .github/actions/notary-server-ip/action.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+name: get-ips-of-service-containers
+description: 'Get alerting endpoint and notary IPs'
```
Comment (Starkteetje) [inline .github/actions/notary-server-ip/action.yml:2]: alerting endpoint doesn't seem to be part of this

File: .github/actions/notary-server-ip/action.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+outputs:
+  notary_ip:
+    description: "IP address of the notary instance"
+    value: ${{ steps.get_notary_server_ip.outputs.notary_ip }}
+runs:
+  using: "composite"
+  steps:
+    - name: Get  IP
```
Comment (Starkteetje) [inline .github/actions/notary-server-ip/action.yml:10]:
```suggestion
    - name: Get IP
```

File: .github/actions/notary-server-ip/action.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      id: get_notary_server_ip
+      run: |
+        NOTARY_CONTAINER=$(docker container ls --no-trunc --format "{{json . }}" | jq ' . | select(.Image|match("notary:server"))')
+        NOTARY_CONTAINER_ID=$(echo ${NOTARY_CONTAINER} | jq -r .ID)
+        NOTARY_CONTAINER_NETWORK=$(echo ${NOTARY_CONTAINER} | jq -r .Networks)
+        NOTARY_SEARCH_PATH=.[0].NetworkSettings.Networks.${NOTARY_CONTAINER_NETWORK}.IPAddress
+        NOTARY_IP=$(docker container inspect ${NOTARY_CONTAINER_ID} | jq -r ${NOTARY_SEARCH_PATH})
```
Comment (Starkteetje) [inline .github/actions/notary-server-ip/action.yml:17]: Not sure whether in theory there could be a whitespace in ${NOTARY_SEARCH_PATH}, but if there was, this would break, thus probably 
```suggestion
        NOTARY_IP=$(docker container inspect ${NOTARY_CONTAINER_ID} | jq -r "${NOTARY_SEARCH_PATH}")
```
is just safer

Comment (Starkteetje) [reply .github/actions/notary-server-ip/action.yml:17]: To a lesser extend (due to nature of ID), same for ${NOTARY_CONTAINER_ID}

File: .github/actions/notary-signer-ip/action.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  using: "composite"
+  steps:
+    - name: Get notary signer IP
+      id: get_signer_ip
+      run: |
+        NOTARY_CONTAINER=$(docker container ls --no-trunc --format "{{json . }}" | jq ' . | select(.Image|match("notary:signer"))')
+        NOTARY_CONTAINER_ID=$(echo ${NOTARY_CONTAINER} | jq -r .ID)
+        NOTARY_CONTAINER_NETWORK=$(echo ${NOTARY_CONTAINER} | jq -r .Networks)
+        NOTARY_SEARCH_PATH=.[0].NetworkSettings.Networks.${NOTARY_CONTAINER_NETWORK}.IPAddress
+        NOTARY_IP=$(docker container inspect ${NOTARY_CONTAINER_ID} | jq -r ${NOTARY_SEARCH_PATH})
```
Comment (Starkteetje) [inline .github/actions/notary-signer-ip/action.yaml:17]: see comment for notary server

File: .github/actions/setup-notary/action.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+name: setup-notary-trust-data
+description: 'Initialize trust data in running notary servvice container for existing test image'
```
Comment (Starkteetje) [inline .github/actions/setup-notary/action.yaml:2]:
```suggestion
description: 'Initialize trust data in running notary service container for existing test image'
```

File: .github/actions/setup-notary/action.yaml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+outputs:
+  root_key:
+    description: "Root key of the securesystemsengineering repository in the ephemeral self-hosted notary instance"
+    value: ${{ steps.get_root_key.outputs.root_key }}
+runs:
+  using: "composite"
+  steps:
+    - name: Install notary and docker client
+      run: |
+        sudo apt install docker
```
Comment (Starkteetje) [inline .github/actions/setup-notary/action.yaml:12]: nit: Could do all installs in this and the next step together

File: .github/workflows/.reusable-integration-test.yml [hunk 9]
```diff
@@ -164,6 +164,81 @@ jobs:
...
+          sudo snap install yq
+      - name: Setup notary signer instance
+        run: |
+          docker run -d -p 7899:7899 -v ./tests/data/notary_service_container/signer:/etc/docker/notary-signer/ notary:signer -config=/etc/docker/notary-signer/config.json
+      - name: Get notary signer instance IP
+        id: get_notary_signer_ip
+        uses: ./.github/actions/notary-signer-ip
+      - name: Setup notary server instance
+        run: |
+          echo ${{ steps.get_notary_signer_ip.outputs.notary_signer_ip }}
```
Comment (Starkteetje) [inline .github/workflows/.reusable-integration-test.yml:206]: Is this still debug code?

File: .github/workflows/.reusable-integration-test.yml [hunk 10]
```diff
@@ -164,6 +164,81 @@ jobs:
...
+        uses: ./.github/actions/notary-server-ip
+      - name: Populate notary instance with trust data
+        uses: ./.github/actions/setup-notary
+        id: setup_notary
+        env:
+          NOTARY_IP: ${{ steps.get_notary_server_ip.outputs.notary_ip }}
+      - uses: ./.github/actions/k8s-version-config
+        name: Setup k8s cluster
+        with:
+          k8s-version: v1.25
```
Comment (Starkteetje) [inline .github/workflows/.reusable-integration-test.yml:219]: Arbitrary choice or is that the oldest supported version?

File: docs/CONTRIBUTING.md (hunk 1 above)
Comment (annekebr) [reply docs/CONTRIBUTING.md:55]: ah why not?

Comment (Starkteetje) [reply docs/CONTRIBUTING.md:55]: because the "usual" installation would be make install or helm install ...

Comment (annekebr) [reply docs/CONTRIBUTING.md:55]: But you're fine with the change itself, just the wording right?

Comment (annekebr) [reply docs/CONTRIBUTING.md:55]: will fix :-)

File: .github/workflows/.reusable-integration-test.yml (hunk 10 above)
Comment (annekebr) [reply .github/workflows/.reusable-integration-test.yml:218]: neither nor I'd say :grin: just copy paste from the general `integration-test` job. BUT: Really good point, bc 1.25 is EOL since end of october, so I think we should change, but I don't know which version is best for testing. Opinion?

https://kubernetes.io/releases/patch-releases/#non-active-branch-history

File: connaisseur/validators/notaryv1/notaryv1_validator.py [hunk 11]
```diff
@@ -118,9 +118,13 @@ async def __process_chain_of_trust(
 
         # load all trust data
         t_start = dt.datetime.now()
+
+        # even if a registry is public like e.g. docker.io, it might still
+        # require a token on repository level, specifying the scope (e.g. pull)
+        repo_token = await self.notary.get_auth(session, image)
```
Comment (Starkteetje) [inline connaisseur/validators/notaryv1/notaryv1_validator.py:124]: While this improves the runtime if there's auth, it adds another network call if the notary doesn't have auth as then `get_auth` retrieves the root data and discards it.

Could make this handling of root data special and have `get_root_trust_and_token` (better name welcome), which returns `(root_data, repo_token)`

File: tests/data/notary/hanswurstnotary.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+trustRoots:
+  - name: library
+    key: |
+      -----BEGIN PUBLIC KEY-----
+      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEH+G0yM9CQ8KjN2cF8iHpiTA9Q69q
+      3brrzLkY1kjmRNOs0c2sx2nm8j2hFZRbyaVsd52Mkw0k5WrX+9vBfbjtdQ==
+      -----END PUBLIC KEY-----
+auth:
+  username: hans
+  password: wurst
```
Comment (Starkteetje) [inline tests/data/notary/hanswurstnotary.yaml:32]: nit: missing newline

File: tests/data/notary_service_container/config/client_config.json [hunk 13]
```diff
@@ -0,0 +1,6 @@
+{
+    "remote_server": {
+            "url": "https://notary.server:4443",
+            "root_ca": "../server/ca.crt"
+    }
+}
```
Comment (Starkteetje) [inline tests/data/notary_service_container/config/client_config.json:6]: nit: missing newline

File: tests/data/notary_service_container/server/config.json [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+      "port": "7899",
+      "tls_ca_file": "./ca.crt",
+      "key_algorithm": "ecdsa",
+      "tls_client_cert": "./notary-server.crt",
+      "tls_client_key": "./notary-server.key"
+    },
+    "storage": {
+      "backend": "memory"
+    }
+}
```
Comment (Starkteetje) [inline tests/data/notary_service_container/server/config.json:22]: nit: missing newline

File: tests/data/notary_service_container/server/config.json [hunk 15]
```diff
@@ -0,0 +1,22 @@
+{
+    "server": {
+      "http_addr": ":4443",
+      "tls_key_file": "./notary-server.key",
+      "tls_cert_file": "./notary-server.crt"
+    },
+    "logging": {
+    "level": "debug"
```
Comment (Starkteetje) [inline tests/data/notary_service_container/server/config.json:8]: Is this a concious decision or part of debugging?

File: tests/data/notary_service_container/signer/config.json [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+      "tls_key_file": "./notary-signer.key",
+      "client_ca_file": "./ca.crt"
+    },
+    "logging": {
+      "level": "debug"
+    },
+    "storage": {
+      "backend": "memory"
+    }
+  }
```
Comment (Starkteetje) [inline tests/data/notary_service_container/signer/config.json:14]: nit: missing newline

File: tests/data/notary_service_container/signer/config.json [hunk 17]
```diff
@@ -0,0 +1,14 @@
+{
+    "server": {
+      "grpc_addr": ":7899",
+      "tls_cert_file": "./notary-signer.crt",
+      "tls_key_file": "./notary-signer.key",
+      "client_ca_file": "./ca.crt"
+    },
+    "logging": {
+      "level": "debug"
```
Comment (Starkteetje) [inline tests/data/notary_service_container/signer/config.json:9]: Conscious or from debugging?

File: tests/integration/cases.yaml [hunk 18]
```diff
@@ -77,6 +77,21 @@ test_cases:
     namespace: default
     expected_msg: Defaulting debug container name to debugger-
     expected_result: VALID
+  self-hosted-notary:
+  - id: ru
```
Comment (Starkteetje) [inline tests/integration/cases.yaml:81]:
```suggestion
  - id: shnu
```
?

File: tests/integration/notary_init.sh [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+send -- "0123456789\r"
+expect "Enter passphrase for new targets key with ID*\r"
+send -- "0123456789\r"
+expect "Repeat passphrase for new targets key with ID*\r"
+send -- "0123456789\r"
+expect "Enter passphrase for new snapshot key with ID*\r"
+send -- "0123456789\r"
+expect "Repeat passphrase for new snapshot key with ID*\r"
+send -- "0123456789\r"
+expect eof
```
Comment (Starkteetje) [inline tests/integration/notary_init.sh:16]: nit: missing newline unless expect scripts don't permit them

File: tests/integration/notary_addhash.sh [hunk 20]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/expect
+
+spawn notary addhash docker.io/securesystemsengineering/testimage self-hosted-notary-signed 528 --sha256 [lindex $argv 0] --publish -c ./tests/data/notary_service_container/config/client_config.json -s https://notary.server:4443 -D
+expect "Enter passphrase for targets key with ID*\r"
+send -- "0123456789\r"
+expect "Enter passphrase for snapshot key with ID*\r"
+send -- "0123456789\r"
+expect eof
```
Comment (Starkteetje) [inline tests/integration/notary_addhash.sh:8]: nit: missing newline unless expect scripts don't permit them

File: tests/integration/update_deployment_notary_test.yaml [hunk 21]
```diff
@@ -0,0 +1,4 @@
+      hostAliases:
+        - ip: ${NOTARY_IP}
+          hostnames:
+            - "notary.server"
```
Comment (Starkteetje) [inline tests/integration/update_deployment_notary_test.yaml:4]: nit: missing newline

File: connaisseur/validators/notaryv1/notary.py [hunk 22]
```diff
@@ -89,39 +89,46 @@ def healthy(self):
...
                         "www-authenticate"
                     ]
                 )
-                token = await self.__get_auth_token(session, auth_url)
-                return await self.get_trust_data(session, image, role, token)
+                repo_token = await self.__get_auth_token(session, auth_url)
+                return repo_token
+            elif status == 401:
+                msg = (
+                    "Unable to find auhorization endpoint in"
```
Comment (Starkteetje) [inline connaisseur/validators/notaryv1/notary.py:109]:
```suggestion
                    "Unable to find authorization endpoint in"
```

File: tests/integration/integration-test.sh [hunk 23]
```diff
@@ -357,12 +357,32 @@ update_helm_for_workloads() {
 	rm update
 }
 
+update_for_self_hosted_notary() {
+	envsubst <tests/integration/update-self-hosted-notary.yaml >update
+	yq '. *+ load("ghcr-values")' -i update
+	yq eval-all --inplace 'select(fileIndex == 0) * select(fileIndex == 1)' helm/values.yaml update
+	rm update
+	curl "https://notary.server:4443/v2/docker.io/securesystemsengineering/testimage/_trust/tuf/root.json" > root.json
+    SELF_HOSTED_NOTARY_PUBLIC_ROOT_KEY_ID=$(cat root.json | jq -r .signatures[0].keyid)
```
Comment (Starkteetje) [inline tests/integration/integration-test.sh:366]: nit: This looks like different indentations

File: tests/integration/integration-test.sh [hunk 24]
```diff
@@ -357,12 +357,32 @@ update_helm_for_workloads() {
...
+    cat root.json | jq -r --arg KEYID $SELF_HOSTED_NOTARY_PUBLIC_ROOT_KEY_ID '.signed.keys | .[$KEYID] | .keyval.public' | base64 -d | openssl x509 -pubkey -noout > self_hosted_notary_root_key.pub
+    yq -i '(.application.validators.[] | select(.name == "self-hosted-notary") | .trustRoots.[] | select(.name=="default") | .key) |= load_str("self_hosted_notary_root_key.pub")' helm/values.yaml
+    rm self_hosted_notary_root_key.pub
```
Comment (Starkteetje) [inline tests/integration/integration-test.sh:369]:
```suggestion
    rm self_hosted_notary_root_key.pub root.json
```
?

File: tests/integration/integration-test.sh [hunk 25]
```diff
@@ -357,12 +357,32 @@ update_helm_for_workloads() {
...
+}
+
 enable_alerting() {
 	envsubst <tests/integration/update-alerting.yaml >update
 	yq eval-all --inplace 'select(fileIndex == 0) * select(fileIndex == 1)' helm/values.yaml update
 	rm update
 }
 
+render_notary_host() {
```
Comment (Starkteetje) [inline tests/integration/integration-test.sh:378]: That's the "we can't yq the template, but need to add values there and sed is fragile, so we do this" part? If so, an explaining comment might be nice

File: tests/integration/integration-test.sh [hunk 26]
```diff
@@ -553,6 +578,12 @@ case $1 in
 	helm_repo_install
 	pre_config_int_test
 	;;
+"self-hosted-notary")
+    update_for_self_hosted_notary
```
Comment (Starkteetje) [inline tests/integration/integration-test.sh:582]: Different indentations again

File: tests/integration/update-self-hosted-notary.yaml [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+  - name: allow
+    type: static
+    approve: true
+  - name: deny
+    type: static
+    approve: false
+  - name: default
+    type: notaryv1
+    host: notary.docker.io
+    trustRoots: []
```
Comment (Starkteetje) [inline tests/integration/update-self-hosted-notary.yaml:15]: Is this superfluous?

File: tests/integration/update-self-hosted-notary.yaml [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+      trustRoot: securesystemsengineering-official
+  - pattern: "k8s.gcr.io/*:*"
+    validator: allow
+  - pattern: docker.io/securesystemsengineering/connaisseur:*
+    validator: allow
+  - pattern: "docker.io/securesystemsengineering/testimage:self-hosted-notary-signed"
+    validator: self-hosted-notary
+  features:
+    detectionMode: false
+    automaticChildApproval: true
```
Comment (Starkteetje) [inline tests/integration/update-self-hosted-notary.yaml:73]: That's not necessary I think

File: connaisseur/validators/notaryv1/notary.py [hunk 29]
```diff
@@ -89,39 +89,56 @@ def healthy(self):
         except Exception:
             return False
 
-    async def get_trust_data(
-        self,
-        session: aiohttp.ClientSession,
-        image: Image,
-        role: TUFRole,
-        token: str = None,
+    async def get_auth_and_root_trust_data(
```
Comment (Starkteetje) [inline connaisseur/validators/notaryv1/notary.py:92]: Since this breaks what one would usually program regarding modularity, I'd say a comment would be beneficical here

Comment (Starkteetje) [reply connaisseur/validators/notaryv1/notary.py:92]: And since you return a tuple, which is always harder to understand, supportive type annotations for the output would be nice

File: connaisseur/validators/notaryv1/notary.py [hunk 30]
```diff
@@ -89,39 +89,56 @@ def healthy(self):
...
                 auth_url = self.__parse_auth(
                     {k.lower(): v for k, v in response.headers.items()}[
                         "www-authenticate"
                     ]
                 )
-                token = await self.__get_auth_token(session, auth_url)
-                return await self.get_trust_data(session, image, role, token)
+                repo_token = await self.__get_auth_token(session, auth_url)
+                root_trust_data, _ = await self.get_auth_and_root_trust_data(
+                    session, image, repo_token
```
Comment (Starkteetje) [inline connaisseur/validators/notaryv1/notary.py:112]: If something is wrong with the registry, this could allow infinite loop: 401-> getToken-> getAuth+Root -> 401 -> getToken -> getAuth+Root...., so probably best to keep the `and not token` part of the original if clause

File: connaisseur/validators/notaryv1/notary.py [hunk 31]
```diff
@@ -119,9 +114,38 @@ async def get_trust_data(
...
+            elif status == 401:
+                msg = (
+                    "Unable to find authorization endpoint in"
+                    "'www-authenticate' header for notary {notary_name}."
+                )
+                raise NotFoundException(message=msg, notary_name=self.name)
+            else:
+                response.raise_for_status()
+                data = await response.text()
+                return TrustData(json.loads(data), "root"), None
```
Comment (Starkteetje) [inline connaisseur/validators/notaryv1/notary.py:131]: nit: this conceptually "swallows" the repo_token

If I call get_x_and_token(a,b,token,c,d) I'd expect to be able to update my `token` with the returned token, whereas that's not possible here. That flow will never be hit, as we never call it from a higher level with a token, but potential for bugs during refactoring

Comment (annekebr) [reply connaisseur/validators/notaryv1/notary.py:131]: Hm, but changing the code to reflect your expectation would mean loosing the benefits we wanted to achieve  (no extra network call) or do I get this wrong? Do you think the function name could be more appropriate? If so, do you have suggestions? (`get_root_trust_data_and_auth_token_if_not_present` :grin: )

Comment (Starkteetje) [reply connaisseur/validators/notaryv1/notary.py:131]:
```suggestion
                return TrustData(json.loads(data), "root"), repo_token
```

would be my suggestion. Basically, if you hand in a token, just return it, thus the return value is always a valid token. I don't think this would break any of the things you wrote, while "fixing" the function's "signature"

**Review by Starkteetje — DISMISSED** [review]

**Review by Starkteetje — APPROVED** [review]
